<a href="https://colab.research.google.com/github/JaeYeon33/Deep-Learning-AZ/blob/main/artificial_neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Artificial Neural Network

## Importing the libraries

In [2]:
import numpy as np
import pandas as pd
import tensorflow as tf

In [3]:
tf.__version__

'2.4.1'

## Part 1 - Data Preprocessing

### Importing the dataset

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [44]:
dataset = pd.read_csv('/content/drive/MyDrive/Udemy/Part 8 - Deep Learning/Section 39 - Artificial Neural Network (ANN)/Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

In [13]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


These are all the desicions of the customers to stay or leave in the bank.

In [45]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column


we want to do it on this specific column of the matrix of Features X and therefore, what we only need to replace here is this Y by that specific column of the matrix of features, X to which we want to apply label encoding.

In [46]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(x[:, 2])

- Female : 0
- Male : 1

In [47]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column

In [48]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

In [49]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [50]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

Feature Scaling is absolutely compulsory for deep learning

Whenever you build an Artificial Neural Network, you have to apply feature scaling.

We will take that full scale and we will based it right back in a new code sale just below Fichus killing will pace it here and now instead of selecting some specific indexes here, we'll just take everything.

So I'm just removing all our index selections here.

In [51]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

So the first thing we have to do is obviously to create a variable that will be nothing else than the artificial neural network itself.

In [52]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [53]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [54]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

Remembering the intuition, lecture's that for the activation function of the output layer. While you don't want to have a rectifier activation function but a sigmoid activation function. 

It's because having a sigmoid activation function allows to get not only ultimately the predictions, but even better.

It will give you the probabilities that the binary outcome is one.

So that we will not only et the predictions of whether the customers choose to leave or not the bank.

But we will also have for each customer the probability that the customer leaves the bank.

And all this thanks to that sigmoid activation function.

So you definitely want that sigmoid activation function for the output layer.

In [55]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Part 3 - Training the ANN

### Compiling the ANN

Optimizer here will update the weights through stochastic gradient descent because we're going to choose Adam Optimizer to add the next iteration, hopefully reduce the loss.

So that's why right here we have to choose an optimizer, but also lost function, which is the way to compute the difference between the predictions and the real result.

And then the accuracy, of course, because that's our final evalution metric. 

So as we said, we're going to choose the Adam optimizer.



What you would have to enter if you were doing non binary classification. You know, like, for example, redicting three different categories. 


For binary classification, the laws must be binary cross entropy and for non binary classification, the laws must be categorical. 


Cross entropy and then also, you knwo, entering non binary classification.


When predicting more than two categories. the activation should not be sigmoid. But softmax.



We can actually choose several metrics at the same time. 


Therefore, in order to enter the values of this parameter, we have to enter them in a pair of square brackets.

In [56]:
ann.compile(optimizer = 'adam' , loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

In [57]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5705 - accuracy: 0.7897
Epoch 2/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4882 - accuracy: 0.7982
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4597 - accuracy: 0.8004
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4587 - accuracy: 0.7951
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4551 - accuracy: 0.7915
Epoch 6/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4325 - accuracy: 0.8028
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4430 - accuracy: 0.7906
Epoch 8/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4280 - accuracy: 0.7974
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4325 - accuracy: 0.7918
Epoch 10/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4126 - accura

That means the out of 100 observations, you have 86 correct predictions.

## Part 4 - Marketing the predictions and evaluating the model

### Predicting the result of a single observation

**HomeWork**

Use our ANN model to predict if the customer with the following informations we will leave the bank:

Geography : France

Credict Score : 600

Gender : Male

Age : 40 years old

Tenure : 3 years

Balance : $ 60000

Number of Products : 2

Does this customer have a credict card ? Yes

Is this customer an Active Member : Yes

Estimated Salary : $ 50000

So, should we say goodbye to that customer?

**Solution**

Predict method should be called arnd observations on which the same scaling was applied as in the training.

And since we trained our Artificial Neural Network with the scaled values, the scaled values of the features, the predict method has to be called on to these observations to which the same scaling was applied.

Check if some scaling was applied during the training.

It is always the case for Neural Networks.

And therefore, in the predict method.

We need to scale that single observation.

And the way to do this is, of course, by calling our SC object

Be careful not to fit transform because the fit transform is used to remember, get the mean and standard deviation of the values and training set in order to fit your scalar to the training set.

But then for the test set, we only need to call the transfer method because if we fitted again the scalar, that would cause some information like it.

In [58]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

[[False]]


Therefore, our ANN model predicts that this customer stay the bank!

**Important note 1** : Notice that values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2** : Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns.

That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encode as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [59]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1))

[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

**The computation of the accuracy**

In [60]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1505   90]
 [ 192  213]]


0.859

We had 1505 correct predictions that the customer stays in the bank.

Then we had 213. Correct predictions that the customer leaves the bank.

Then we have 90 incorrect predictions that the customer leaves the bank.

And 192 incorrect predictions that the customer stays in the bank.


Accuracy of 85 percent is a very decent accuracy.